[LoopPeel] Fix branch weights' effect on block frequencies

jdenny-ornl opened the pull request:
[LoopPeel] Fix branch weights' effect on block frequencies

This patch implements the LoopPeel changes discussed in [[RFC] Fix Loop Transformations to Preserve Block Frequencies](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785).

In summary, a loop's latch block can have branch weight metadata that encodes an estimated trip count that is derived from application profile data.  Initially, the loop body's block frequencies agree with the estimated trip count, as expected.  However, sometimes loop transformations adjust those branch weights in a way that correctly maintains the estimated trip count but that corrupts the block frequencies.  This patch addresses that problem in LoopPeel, which it changes to:

- Maintain branch weights consistently with the original loop for the sake of preserving the total frequency of the original loop body.
- Store the new estimated trip count in the `llvm.loop.estimated_trip_count` metadata, introduced by PR #148758.

jdenny-ornl committed:
[LoopPeel] Fix branch weights' effect on block frequencies

For example:

```
declare void @f(i32)

define void @test(i32 %n) {
entry:
  br label %do.body

do.body:
  %i = phi i32 [ 0, %entry ], [ %inc, %do.body ]
  %inc = add i32 %i, 1
  call void @f(i32 %i)
  %c = icmp sge i32 %inc, %n
  br i1 %c, label %do.end, label %do.body, !prof !0

do.end:
  ret void
}

!0 = !{!"branch_weights", i32 1, i32 9}
```

Given those branch weights, once any loop iteration is actually
reached, the probability of the loop exiting at the iteration's end is
1/(1+9).  That is, the loop is likely to exit every 10 iterations and
thus has an estimated trip count of 10.  `opt
-passes='print<block-freq>'` shows that 10 is indeed the frequency of
the loop body:

```
Printing analysis results of BFI for function 'test':
block-frequency-info: test
 - entry: float = 1.0, int = 1801439852625920
 - do.body: float = 10.0, int = 18014398509481984
 - do.end: float = 1.0, int = 1801439852625920
```

Key Observation: The frequency of reaching any particular iteration is
less than for the previous iteration because the previous iteration
has a non-zero probability of exiting the loop.  This observation
holds even though every loop iteration, once actually reached, has
exactly the same probability of exiting and thus exactly the same
branch weights.

Now we use `opt -unroll-force-peel-count=2 -passes=loop-unroll` to
peel 2 iterations and insert them before the remaining loop.  We
expect the key observation above not to change, but it does under the
implementation without this patch.  The block frequency becomes 1.0
for the first iteration, 0.9 for the second, and 6.4 for the main loop
body.  Again, a decreasing frequency is expected, but it decreases too
much: the total frequency of the original loop body becomes 8.3.  The
new branch weights reveal the problem:

```
!0 = !{!"branch_weights", i32 1, i32 9}
!1 = !{!"branch_weights", i32 1, i32 8}
!2 = !{!"branch_weights", i32 1, i32 7}
```

The exit probability is now 1/10 for the first peeled iteration, 1/9
for the second, and 1/8 for the remaining loop iterations.  It seems
this behavior is trying to ensure a decreasing block frequency.
However, as in the key observation above for the original loop, that
happens correctly without decreasing the branch weights across
iterations.

This patch changes the peeling implementation not to decrease the
branch weights across loop iterations so that the frequency for every
iteration is the same as it was in the original loop.  The total
frequency of the loop body, summed across all its occurrences, thus
remains 10 after peeling.

Unfortunately, that change means a later analysis cannot accurately
estimate the trip count of the remaining loop while examining the
remaining loop in isolation without considering the probability of
actually reaching it.  For that purpose, this patch stores the new
trip count as separate metadata named `llvm.loop.estimated_trip_count`
and extends `llvm::getLoopEstimatedTripCount` to prefer it, if
present, over branch weights.

An alternative fix is for `llvm::getLoopEstimatedTripCount` to
subtract the `llvm.loop.peeled.count` metadata from the trip count
estimated by a loop's branch weights.  However, there might be other
loop transformations that still corrupt block frequencies in a similar
manner and require a similar fix.  `llvm.loop.estimated_trip_count` is
intended to provide a general way to store estimated trip counts when
branch weights cannot directly store them.

This patch introduces several fixme comments that need to be addressed
before it can land.

jdenny-ornl committed:
Run update_test_checks.py on a test

jdenny-ornl committed:
Fix typo

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Document new metadata

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -7866,6 +7866,17 @@ The attributes in this metadata is added to all followup loops of the
 
+'``llvm.loop.estimated_trip_count``' Metadata
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Should explain where it should be attached. What if it's not placed on an arbitrary terminator that isn't a loop backed?

jdenny-ornl replied:
> Should explain where it should be attached. What if it's not placed on an arbitrary terminator that isn't a loop backed?

The [`llvm.loop` metadata description](https://llvm.org/docs/LangRef.html#llvm-loop) explains placement generally for all `llvm.loop.*` metadata.  I've also added an example in the same manner as the docs I see for other `llvm.loop.*` metadata.

> Consequences for it being incorrect?

I've added a description of how it's consumed.

jdenny-ornl replied:
This is now part of PR #148758.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -7866,6 +7866,17 @@ The attributes in this metadata is added to all followup loops of the
+'``llvm.loop.estimated_trip_count``' Metadata
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This metadata records the loop's estimated trip count.  If it is not present, a
+loop's estimated trip count should be computed from any ``branch_weights``
+metadata attached to the latch block's branch instruction.
+
+Thus, this metadata frees loop transformations to compute latch branch weights
+solely for the purpose of maintaining accurate block frequencies instead of
+requiring the branch weights to always serve both roles.
```
This reads as more background information rather than a direct description of the meaning of the metadata

jdenny-ornl replied:
I've added some text about the semantics.

jdenny-ornl replied:
This is now part of PR #148758.

jdenny-ornl commented on llvm/docs/LangRef.rst:
```diff
@@ -7866,6 +7866,17 @@ The attributes in this metadata is added to all followup loops of the
 loop distribution pass. See
 :ref:`Transformation Metadata <transformation-metadata>` for details.
 
+'``llvm.loop.estimated_trip_count``' Metadata
```
In the other loop metadata, I understood `.` to separate levels in a naming hierarchy while `_` separates words in a single level's name.

For example: [‘llvm.loop.unroll_and_jam.count’ Metadata](https://llvm.org/docs/LangRef.html#llvm-loop-unroll-and-jam-count-metadata)

I didn't think of any way to break `estimated_trip_count` apart into a meaningful hierarchy.

Thanks for reviewing.  I'll work on your other comments.

jdenny-ornl replied:
This is now part of PR #148758.

jdenny-ornl committed:
Improve LangRef.rst entry

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Update fixmes

Extending beyond the limitations of `getExpectedExitLoopLatchBranch`
is a possible improvement for the future not an urgent fixme.

No one has pointed out code that computes estimated trip counts
without using `llvm::getLoopEstimatedTripCount`.

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

MatzeB commented:
- My understanding is that this fixes the branch weights (basically by removing all the code making adjustments which turned out to preserve the trip count estimation but at the price of wrong branch weights) as discussed in the RFC. I think this part is great!
- I'm not a fan of introducing the new metadata as I'm not convinced we even need it / have users for it; but if there was indeed use cases that required the loop trip count as a separate metric, then I guess I can live with the new metadata, just wanted to make sure we actually need it.

mtrofin commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -850,27 +852,35 @@ llvm::getLoopEstimatedTripCount(Loop *L,
             getEstimatedTripCount(LatchBranch, L, ExitWeight)) {
       if (EstimatedLoopInvocationWeight)
         *EstimatedLoopInvocationWeight = ExitWeight;
+      if (auto EstimatedTripCount =
```
When would `LLVMLoopEstimatedTripCount` not be present, i.e. shouldn't one expect that, if the build is (for example) using instrumented profiling, the metadata should be present, unless something dropped it, which should be a bug?

jdenny-ornl replied:
Currently in this PR, the new metadata is not introduced until a pass like LoopPeel calls setLoopEstimatedTripCount.  Normally that would be because it creates the situation where branch weights alone cannot encode both block frequencies and estimated trip counts, so the new metadata is then required (but of course not all passes have been fixed yet to maintain block frequencies correctly).

Moreover, the todo comment this PR introduces into LoopPeel.cpp describes how getLoopEstimatedTripCount and setLoopEstimatedTripCount cannot handle some loop forms and so don't get/set the estimated trip count.  That issue exists without this PR.  What's new is that it skips adding the new metadata in that case.

As we discussed at some point in the RFC, the inconsistent presence of the new metadata is not ideal, and we might want to ultimately change that.  But in the long term, it sounds like people might want to drop estimated trip counts altogether, so I'm not yet convinced that efforts to make the new metadata more consistently present are worthwhile.  What do you think?

jdenny-ornl replied:
Surprises should be avoidable if we can assume accesses to the new metadata will always be via the existing get/set methods.  Perhaps that's too big of an assumption, and it at least suggests more documentation encouraging that constraint.

But an "unknown" state does sound like a reasonable solution.  I'll look into implementing that.  Thanks.

jdenny-ornl committed:
Update test for AArch4, which I did not build before

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Run update script on test changed by merge from main

The update adds this PR's new metadata.

jdenny-ornl committed:
[PGO] Add `llvm.loop.estimated_trip_count` metadata

This patch implements the `llvm.loop.estimated_trip_count` metadata
discussed in [[RFC] Fix Loop Transformations to Preserve Block
Frequencies](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785).
As [suggested in the RFC
comments](https://discourse.llvm.org/t/rfc-fix-loop-transformations-to-preserve-block-frequencies/85785/4),
it adds the new metadata to all loops at the time of profile ingestion
and estimates each trip count from the loop's `branch_weights`
metadata.  As [suggested in the PR#128785
review](https://github.com/llvm/llvm-project/pull/128785#discussion_r2151091036),
it does so via a `PGOEstimateTripCountsPass` pass, which creates the
new metadata for the loop but omits the value if it cannot estimate a
trip count due to the loop's form.

An important observation not previously discussed is that
`PGOEstimateTripCountsPass` *often* cannot estimate a loop's trip
count but later passes can transform the loop in a way that makes it
possible.  Currently, such passes do not necessarily update the
metadata, but eventually that should be fixed.  Until then, if the new
metadata has no value, `llvm::getLoopEstimatedTripCount` disregards it
and tries again to estimate the trip count from the loop's
`branch_weights` metadata.

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Add PGOEstimateTripCounts in more cases

jdenny-ornl committed:
Add unused initialization

jdenny-ornl committed:
Simplify some test changes

jdenny-ornl committed:
Extend verify pass to cover new metadata

jdenny-ornl committed:
Fix test for some builds

Somehow, on some of my builds, `llvm::` prefixes are dropped from some
symbol names in the printed past list.

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Apply some small reviewer suggestions

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Update for merge from pgo-estimated-trip-count

That's PR #128785, which is now a parent PR.

First, remove a todo that's now documented more generally than
LoopPeel in plenty of other places.  Second, update LoopPeel's
setLoopEstimatedTripCount call to avoid a now redundant argument that
eventually won't be supported.

jdenny-ornl committed:
Attempt to fix windows pre-commit CI

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Run update script on new test from last merge

For estimated trip count metdata.

jdenny-ornl committed:
Reapply 3a18fe33f0763cd9276c99c276448412100f6270

That was applied after merging this PR last time.

jdenny-ornl committed:
Convert to function pass, avoid needless pass invalidation

Suggested at
<https://github.com/llvm/llvm-project/pull/148758#discussion_r2246193446>.

jdenny-ornl committed:
Fix layering violation

Suggested at
<https://github.com/llvm/llvm-project/pull/148758#issuecomment-3141478969>.

jdenny-ornl committed:
Apply clang-format

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Remove PGOEstimateTripCountsPass and no-value form of metadata

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Fix case where nested loops share latch

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Remove redundant code

jdenny-ornl committed:
Clarify recent comments some

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into pgo-estimated-trip-count

jdenny-ornl committed:
Merge branch 'pgo-estimated-trip-count' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl commented:
> @[jdenny-ornl](https://github.com/jdenny-ornl) jdenny-ornl requested review from [aaupov](https://github.com/aaupov), [maksfb](https://github.com/maksfb), [rafaelauler](https://github.com/rafaelauler), [ayermolo](https://github.com/ayermolo), [yota9](https://github.com/yota9), [paschalis-mpeis](https://github.com/paschalis-mpeis), [yozhu](https://github.com/yozhu), [JDevlieghere](https://github.com/JDevlieghere), [lanza](https://github.com/lanza), [bcardosolopes](https://github.com/bcardosolopes), [xlauko](https://github.com/xlauko), [andykaylor](https://github.com/andykaylor), https://github.com/orgs/llvm/teams/reviewers-libunwind, https://github.com/orgs/llvm/teams/reviewers-libcxxabi and https://github.com/orgs/llvm/teams/reviewers-libcxx as [code owners](https://github.com/llvm/llvm-project/blob/b8aed9bf991b7cd4ed8b07bc32537d7ece6bb7c2/.github/CODEOWNERS#L136)

No, I didn't.

jdenny-ornl commented:
> [Release Binaries All / Build Release Binaries (ubuntu-22.04-arm) / Build Release Package (pull_request)](https://github.com/llvm/llvm-project/actions/runs/17739363757/job/50409434332?pr=128785)S

I think CI has gone awry.

jdenny-ornl committed:
Empty commit to try to restart pre-commit CI

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl committed:
Merge branch 'main' into fix-peel-branch-weights

jdenny-ornl commented on llvm/lib/Transforms/Utils/LoopPeel.cpp:
```diff
@@ -1426,15 +1338,52 @@ bool llvm::peelLoop(Loop *L, unsigned PeelCount, bool PeelLast, LoopInfo *LI,
+  // while examining it in isolation without considering the probability of
+  // actually reaching it, we store the new trip count as separate metadata.
+  if (auto EstimatedTripCount = getLoopEstimatedTripCount(L)) {
+    // FIXME: The previous updateBranchWeights implementation had this
+    // comment:
+    //
+    //   Don't set the probability of taking the edge from latch to loop header
+    //   to less than 1:1 ratio (meaning Weight should not be lower than
+    //   SubWeight), as this could significantly reduce the loop's hotness,
+    //   which would be incorrect in the case of underestimating the trip count.
+    //
+    // See e8d5db206c2f commit log for further discussion.  That seems to
+    // suggest that we should avoid ever setting a trip count of < 2 here
+    // (equal chance of continuing and exiting means the loop will likely
+    // continue once and then exit once).  Or is keeping the original branch
+    // weights already a sufficient improvement for whatever analysis cares
+    // about this case?
```
After many months, there have been no remarks on this fixme comment.  I am inclined to just remove the comment.

@mtrofin Thanks for accepting.  Is it ok with you if I just remove it?

jdenny-ornl replied:
I have decided to move forward with removing the fixme and merging the PR.  If there are objections after the merge, I can revert or fix.

jdenny-ornl committed:
Remove fixme about probability < 1:1 ratio

I added the fixme to this PR because I do not understand the
importance of the associated prior comments from others.  There has
been no response to the fixme since I posted the PR seven months ago,
so I assume it is not important going forward.
